src/V3Param.cpp

```diff
@@ -63,908 +64,980 @@
+        // Collect constants from original module for not overridden parameters, so potentially
+        // don't need to clone a new module
+        bool hasEmptyParam = false;
+        for (size_t i = 0; i < modInfop->paramList().size(); i++) {
+            const AstNode* const paramp = modInfop->paramListAt(i);
+            if (collectedParams->m_params[i]) continue;
+            if (const AstVar* const varp = VN_CAST(paramp, Var)) {
+                if (!varp->valuep()) {  // skip parameter with no value
+                } else if (AstConst* constp = VN_CAST(varp->valuep(), Const)) {
+                    if (isString(varp->subDTypep())) {
+                        constp = VN_AS(replaceWithStringp(constp), Const);
                     }
+                    if (modInfop->hierBlock() && constp->isOpaque()) continue;
+                    collectedParams->m_params[i] = constp;
+                } else {
+                    if (modInfop->hierBlock()) continue;
+                    hasEmptyParam = true;
+                }
+            } else if (const AstParamTypeDType* const dtypep = VN_CAST(paramp, ParamTypeDType)) {
+                if (modInfop->hierBlock()) continue;
+                if (!dtypep->childDTypep()) {  // skip parameter with no value
+                } else {
+                    // We are not sure if item.first->childDTypep() is constant (e.g. has
+                    // non-constant range), so just mark it currently
+                    hasEmptyParam = true;
                 }
             }
         }
-    }
-
-    void storeOriginalParams(AstClass* const classp) {
-        for (AstNode* stmtp = classp->stmtsp(); stmtp; stmtp = stmtp->nextp()) {
-            AstNode* originalParamp = nullptr;
-            if (AstVar* const varp = VN_CAST(stmtp, Var)) {
-                if (varp->isGParam()) originalParamp = varp->clonep();
-            } else if (AstParamTypeDType* const ptp = VN_CAST(stmtp, ParamTypeDType)) {
-                if (ptp->isGParam()) originalParamp = ptp->clonep();
+        (void)hasEmptyParam;
+        collectedParams->rehash();
+        foundp = modInfop->findNodeWithOverriddenParamSet(collectedParams.get());
+        if (!foundp) foundp = modInfop->findNodeWithFullParamSet(collectedParams.get());
+        if (foundp) {
+            UINFO(7, "  module found with collected param set\n");
+            modInfop->insertOverriddenParamSet(overriddenParams.release(), foundp);
+            modInfop->insertOverriddenParamSet(collectedParams.release(), foundp);
+            return foundp;
+        }
+        AstNodeModule* clonedModp = nullptr;
+        // Even if there's no unknown parameters, we still have to re-evaluate them under module's
+        // context (to perform truncation/extension/type propagation/...)
+        auto evaluatedParams = std::make_unique<ModParamSet>(*overriddenParams);
+        clonedModp = evaluateModParams(modp, modInfop, evaluatedParams.get());
```

Comment:
No matter whether the cell has overridden parameters or not it always make a new copy (in deepCloneModule called by evaluateModParams), if there're no previously created instance (for both modules and classes), because we don't know the default values so can't determine whether it does 'override'. It the situation mentioned [here](https://github.com/verilator/verilator/pull/4581#issuecomment-1770956387).